scale-decode-derive/src/lib.rs

```diff
@@ -0,0 +1,563 @@
+
+    // determine what the body of our visitor functions will be based on the type of enum fields
+    // that we're trying to generate output for.
+    let variant_ifs = details.variants.iter().map(|variant| {
+        let variant_ident = &variant.ident;
+        let variant_name = variant_ident.to_string();
+
+        let visit_one_variant_body = match &variant.fields {
+            syn::Fields::Named(fields) => {
+                let (
+                    field_count,
+                    field_composite_keyvals,
+                    field_tuple_keyvals
+                ) = named_field_keyvals(path_to_scale_decode, fields);
+
+                quote!{
+                    let fields = value.fields();
+                    return if fields.has_unnamed_fields() {
+                        if fields.remaining() != #field_count {
+                            return Err(#path_to_scale_decode::Error::new(#path_to_scale_decode::error::ErrorKind::WrongLength {
+                                actual_len: fields.remaining(),
+                                expected_len: #field_count
+                            }));
+                        }
+                        let vals = fields;
+                        Ok(#path_to_type::#variant_ident { #(#field_tuple_keyvals),* })
+                    } else {
+                        let vals: ::std::collections::HashMap<Option<&str>, _> = fields
+                            .map(|res| res.map(|item| (item.name(), item)))
+                            .collect::<Result<_, _>>()?;
+                        Ok(#path_to_type::#variant_ident { #(#field_composite_keyvals),* })
+                    }
+                }
+            },
+            syn::Fields::Unnamed(fields) => {
+                let (
+                    field_count,
+                    field_vals
+                ) = unnamed_field_vals(path_to_scale_decode, fields);
+
+                quote!{
+                    let fields = value.fields();
+                    if fields.remaining() != #field_count {
+                        return Err(#path_to_scale_decode::Error::new(#path_to_scale_decode::error::ErrorKind::WrongLength {
+                            actual_len: fields.remaining(),
+                            expected_len: #field_count
+                        }));
+                    }
+                    let vals = fields;
+                    return Ok(#path_to_type::#variant_ident ( #(#field_vals),* ))
+                }
+            },
+            syn::Fields::Unit => {
+                quote!{
+                    return Ok(#path_to_type::#variant_ident)
+                }
+            },
+        };
+
+        quote!{
+            if value.name() == #variant_name {
+                #visit_one_variant_body
+            }
+        }
+    });
+
+    quote!(
+        #[doc(hidden)]
+        #visibility struct #visitor_struct_name #impl_generics (
+            ::std::marker::PhantomData<#phantomdata_type>
+        );
+
```

Comment:
nit: Should we place these impls under `const _: () = {` similar to how we generate the code for structs?

Reply:
Good idea; I was actually thinking about exactly this at some point earlier and then forgot (it would be nice to hide the struct inparticular while avoiding any "must export Visitor because it's part of a public interface" things.

Lemme have a look!

Reply:
I put everyting behind const, and hid the Visitor structs behind the `const _: () = {}` thing too (quite a neat trick that is), which also meant I no logner needed to generate any unique names for them and such :)